src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,710 @@
+
+    return true;
+}
+
+static inline uint32_t
+getRecordSize(const unsigned char *const buffer)
+{
+    return ((XLogRecord *) (buffer))->xl_tot_len;
+}
+
+// Returns ReadRecordSuccess on success record read and returns ReadRecordNeedBuffer if a new input buffer is needed to continue
+// reading.
+static ReadRecordStatus
+readRecord(WalFilterState *const this, const Buffer *const input)
+{
+    // Go back to the place where the reading was interrupted due to the exhaustion of the input buffer.
+    switch (this->currentStep)
+    {
+        case noStep:
+            // noting to do
+            break;
+
+        case stepBeginOfRecord:
+            goto stepBeginOfRecord;
+
+        case stepReadHeader:
+            goto stepReadHeader;
+
+        case stepReadBody:
+            goto stepReadBody;
+    }
+
+    if (this->pageOffset == 0 || this->pageOffset >= this->walPageSize)
+    {
+stepBeginOfRecord:
+        if (!readPage(this, input, stepBeginOfRecord))
+        {
+            return ReadRecordNeedBuffer;
+        }
+
+        if (this->isBegin)
+        {
+            this->isBegin = false;
+            // There may be an unfinished record from the previous file at the beginning of the file. Just skip it.
+            // We should have handled it elsewhere.
+            if (this->currentHeader->xlp_info & XLP_FIRST_IS_CONTRECORD &&
+                !(this->currentHeader->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD))
+            {
+                this->pageOffset += MAXALIGN(this->currentHeader->xlp_rem_len);
+            }
+        }
+        else if (this->currentHeader->xlp_info & XLP_FIRST_IS_CONTRECORD)
+        {
+            THROW_FMT(FormatError, "%s - should not be XLP_FIRST_IS_CONTRECORD", strZ(pgLsnToStr(this->recPtr)));
+        }
+    }
+
+    // Record header can be split between pages but first field xl_tot_len is always on single page
+    uint32_t record_size = getRecordSize(this->page + this->pageOffset);
+
+    if (this->recBufSize < record_size)
+    {
+        MEM_CONTEXT_OBJ_BEGIN(this)
+        {
+            this->record = memResize(this->record, record_size);
+        }
+        MEM_CONTEXT_OBJ_END();
+        this->recBufSize = record_size;
+    }
+
+    memcpy(this->record, this->page + this->pageOffset, Min(SizeOfXLogRecord, this->walPageSize - this->pageOffset));
+
+    this->totLen = this->record->xl_tot_len;
+
+    // If header is split read rest of the header from next page
+    if (SizeOfXLogRecord > this->walPageSize - this->pageOffset)
+    {
+        this->gotLen = this->walPageSize - this->pageOffset;
+stepReadHeader:
+        if (!readPage(this, input, stepReadHeader))
+        {
+            return ReadRecordNeedBuffer;
+        }
+
+        if (this->currentHeader->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD)
+        {
+            // This record has been overwritten.
+            // Write to the output what we managed to read as is, skipping filtering.
+            return ReadRecordSuccess;
```

Comment:
If in this case the record is not read completely, then what will be written next?

Reply:
We write to the output as much as we have read. Based on the `this->gotLen` field.